### ISSUE TITLE
Confusion (bug?) about setting linkType in link parsing

### ISSUE THREAD

**Issue by hartzell** [russross/blackfriday#441]
I'm confused about the code that sets the linkType in `link()`.

It seems like the [case at 216](https://github.com/russross/blackfriday/blob/master/inline.go#L216) and [the else clause in the case at 223](https://github.com/russross/blackfriday/blob/master/inline.go#L223) are redundant.

I'm wondering if the [case at 216](https://github.com/russross/blackfriday/blob/master/inline.go#L216) was also meant to also be checking `data[offset-1] == '!'` like [the case at 219](https://github.com/russross/blackfriday/blob/master/inline.go#L219) does?

**Comment by rtfb** [comment]
Yes, it does seem that the else clause on line 226 can be removed, because its conditions are duplicate to the ones of the case at 216. I _think_ that additional check for `'!'` is not necessary in the case at 216, and the comment above is meant to inform that a check for "footnotes enabled && caret follows open square" must precede a check for `"![image link]"` in order to avoid interpreting it as `"![^image link]"`. I hope this explanation is a bit more readable than the code... it's gnarly :-)

BTW, I reformatted that code in c5943e068510, supposedly after a session of staring into this code for a while, but apparently, it didn't become readable enough :-). Feel free to clean it up further.

**Comment by hartzell** [comment]
I'm going to leave it as is for the moment, keep the PR simpler and cleaner.  It's enough for me to know that I'm not obviously missing the point.  If I get a round tuit (around to it...) I might take a swipe at it but if someone else wants to tidy it up I won't complain.
